File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    
    remains = [-w for w in works]
    heapq.heapify(remains)
    
    for _ in range(n):
        if not remains:
            break
        
        temp = -heapq.heappop(remains)
        if temp <= 0:
            break
        temp -= 1
        heapq.heappush(remains, -temp)
    
    
    return sum((x ** 2) for x in remains)  

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        
        PriorityQueue<Integer> q = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int work : works) {
            q.offer(work);
        }
        
        for (int i = 0; i < n; i++) {
            if (q.isEmpty()) break;
            
            int temp = q.poll();
            if (temp <= 0) break;
            
            q.offer(temp - 1);
        }
        
        for (int remain : q) {
            answer += (long) remain * remain;
        }
        
        return answer;
    }
}
